### ISSUE TITLE
Fix TestScript/gopath_std_vendor test in microsoft/dev.boringcrypto.go1.18: our additional vendored library interferes

### ISSUE THREAD

**Issue by dagood** [microsoft/go#481]
https://github.com/microsoft/go/pull/479 hit an issue with `TestScript/gopath_std_vendor` when trying to update to 1.18.0. I'm filing an issue so it's easier to refer to in the future rather than being buried in a PR discussion thread.

The test failure repros locally with `../bin/go test ./cmd/go -run TestScript/gopath_std_vendor` in `go/src` after building:

https://gist.github.com/dagood/98ecf0e1656061bdb3d5f87ab22cab50

```
[...]
            > go list -test -deps -f '{{.ImportPath}} {{.Dir}}' .
            [stdout]
            vendor/golang.org/x/net/http2/hpack $WORK/gopath/src/vendor/golang.org/x/net/http2/hpack
            issue16333 $WORK/gopath/src/issue16333
            internal/goarch /home/dagood/git/go/go/src/internal/goarch
[...]
            vendor/github.com/microsoft/go-crypto-openssl/openssl/internal/subtle /home/dagood/git/go/go/src/vendor/github.com/microsoft/go-crypto-openssl/openssl/internal/subtle
            runtime/cgo /home/dagood/git/go/go/src/runtime/cgo
            vendor/github.com/microsoft/go-crypto-openssl/openssl /home/dagood/git/go/go/src/vendor/github.com/microsoft/go-crypto-openssl/openssl
            crypto/internal/backend /home/dagood/git/go/go/src/crypto/internal/backend
[...]
            > stdout $GOPATH[/\\]src[/\\]vendor[/\\]golang.org[/\\]x[/\\]net[/\\]http2[/\\]hpack
            > ! stdout $GOROOT[/\\]src[/\\]vendor
            FAIL: testdata/script/gopath_std_vendor.txt:28: unexpected match for `(?m)/home/dagood/git/go/go[/\\]src[/\\]vendor` found in stdout: /home/dagood/git/go/go/src/vendor
```

So, why is this showing up now in 1.18, not 1.17? The test script hasn't changed for three years.

I tried [the `-work` command to keep the test files around](https://github.com/golang/go/blob/master/src/cmd/go/testdata/script/README), but it doesn't seem to work--under `$WORK` I just see a few hundred numbered folders with some unrelated-seeming files in them.

Instead, I added `go list -test -deps -f '{{.Dir}} imports: {{ .Imports }}'` to the test script directly to get the test to print some more info. With that, I was able to work backwards to find this dependency chain that brings go-crypto-openssl into the dependencies of the test program:

```
.../go/src/internal/fuzz imports: [bytes context crypto/sha256 encoding/binary encoding/json errors fmt go/ast go/parser go/token internal/godebug internal/unsafeheader io io/ioutil math math/bits os os/exec path/filepath reflect runtime strconv strings sync sync/atomic syscall time unicode/utf8 unsafe]
.../go/src/crypto/sha256 imports: [crypto crypto/internal/backend encoding/binary errors hash internal/cpu math/bits]
.../go/src/crypto/internal/backend imports: [crypto/internal/boring/sig vendor/github.com/microsoft/go-crypto-openssl/openssl syscall]
.../go/src/vendor/github.com/microsoft/go-crypto-openssl/openssl imports: [C crypto crypto/cipher crypto/subtle encoding/asn1 errors vendor/github.com/microsoft/go-crypto-openssl/openssl/internal/subtle hash math/big runtime strconv strings unsafe]
```

Fuzz testing is new to go1.18, so that explains it!

The solution I'm trying now is to exclude our library from the stdout check. The test is only *trying* to use the vendored hpack library. go-crypto-openssl is a bystander. I don't fully understand https://github.com/golang/go/issues/16333, but this doesn't seem to be a "make sure everything is vendored properly" test--more about making sure a weird interaction between a vendored standard library and the same library already existing in GOPATH doesn't regress.

Since this test file hasn't changed in a long time, hopefully the patch won't require much maintenance if that ends up being the long-term way to fix this.

/cc @microsoft/golang-compiler

**Comment by dagood** [comment]
I added a patch to https://github.com/microsoft/go/pull/479 to skip the test altogether. It seems preferable to adding a subtle regex, and it'll be even less likely to need maintenance over time. I doubt we will ever break something that this test would have caught. (Some subtle vendoring, import, and build behavior.)

**Comment by qmuntal** [comment]
Good job tracking down the failure. Skipping this particular test is a good solution, yet my philosophy is to avoid skipping upstream test by all means, so here is another approach:

The OpenSSL backend is not used *and* excluded from the compilation when the build tag `gocrypto` is used. If we change the `go build .` invocation in `gopath_std_vendor.txt` to `go build -tags=gocrypto .`, then the vendored openssl package won't appear in the dependency list.

If doing it this way I would add the fix into `0003-Integrate-OpenSSL-crypto-module.patch` instead of creating a new one, as that patch is already changing some tests to accommodate the new imports.

**Comment by dagood** [comment]
Aha, nice, I'll use that instead and rebase the patch file. In IM you mentioned that it might be good to move all test adjustments into a new patch, I'll do that too.

**Comment by qmuntal** [comment]
@dagood this issue was fixed in #516. Can we close it?

**Comment by dagood** [comment]
Whoops, yep! Thanks, closing.
